SCP-1733 Marlowe Playground Frontend: Refactor the bottom panel into a reusable component

hrajchert opened the pull request:
This PR refactors the bottom panel functionality that was shared through copy and paste in the HaskellEditor, JavaScriptEditor, MarloweEditor and Simulator pages into a single reusable component.

This way, the functionalities are reused and they share visual aspects. Before this PR for example, there was no way of "switching tabs" between generated code and code errors in the Javascript and Haskell editor. Granted, there can never coexist at the same time, but, they also didn't have a title that indicated what were the possible tabs.

We should take special attention to the types described in `BottomPanel.Types`, which defines the first use case we had for a sub-component to be able to display HTML and fire actions inside the sub-component but the actions being defined in the parent. Kind of how the `transclude` functionality used to work in Angular 1.

## Haskell editor before 
* Without errors or compiled output
<img width="1789" alt="Screen Shot 2021-01-21 at 20 09 29" src="https://user-images.githubusercontent.com/2634059/105423912-21782780-5c25-11eb-97d2-158c5b381769.png">
* With compiled output
<img width="1773" alt="Screen Shot 2021-01-21 at 20 07 30" src="https://user-images.githubusercontent.com/2634059/105423943-2e951680-5c25-11eb-8220-8e68d8d23f11.png">
* With errors
<img width="1785" alt="Screen Shot 2021-01-21 at 20 07 47" src="https://user-images.githubusercontent.com/2634059/105423974-394fab80-5c25-11eb-95bf-93c48412075b.png">

## Haskell editor with PR
* Without errors or compiled output
<img width="1788" alt="Screen Shot 2021-01-21 at 20 08 38" src="https://user-images.githubusercontent.com/2634059/105424036-52f0f300-5c25-11eb-9db7-4359ebb8119f.png">

* With compiled output
<img width="1784" alt="Screen Shot 2021-01-21 at 20 08 08" src="https://user-images.githubusercontent.com/2634059/105424059-5c7a5b00-5c25-11eb-8f40-05897ba3a8cf.png">

* With errors
<img width="1790" alt="Screen Shot 2021-01-21 at 20 08 27" src="https://user-images.githubusercontent.com/2634059/105424073-6603c300-5c25-11eb-97b1-e04f4d2289d3.png">

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A plutus.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [ ] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [x] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [x] Someone approved it
- [x] Commits have useful messages
- [x] Review clarifications made it into the code
- [x] History is moderately tidy; or going to squash-merge

shmish111 commented on marlowe-playground-client/src/BottomPanel/State.purs:
```diff
@@ -0,0 +1,21 @@
+
+handleAction ::
+  forall m panel panelAction.
+  Action panel panelAction ->
```
I feel like `panelAction` should be `action`. It the action type of the "parent"

hrajchert replied:
changed

shmish111 commented on marlowe-playground-client/src/BottomPanel/Types.purs:
```diff
@@ -0,0 +1,44 @@
+-- we have, and the panelAction, which is needed to allow the contents of the panel, to fire actions
+-- on the parent. The `panel` type could eventually be refactored into just having a Map/Record, if we dont
+-- care for it to be type safe, and the PanelAction could eventually be refactored to Variants/Run.
+data Action panel panelAction
```
could you explain a bit more about what a `panel` is in the specific context of this UI element. I.e. "the bottom panel has different tabs, panel defines which tab is selected" or something

hrajchert replied:
renamed panelAction and added better explanation

shmish111 commented on marlowe-playground-client/src/BottomPanel/View.purs:
```diff
@@ -0,0 +1,65 @@
+  (panel -> HTML p (Action panel panelAction)) ->
+  State panel ->
+  HTML p (Action panel panelAction)
+render panelTitles panelContents state =
```
a comment with some explanation of the parameters (e.g. panelTitles is an ordered list of the tab titles)

hrajchert replied:
Added some comments

shmish111 commented on marlowe-playground-client/src/SimulationPage/View.purs:
```diff
@@ -40,15 +44,28 @@ render ::
+
+  currentStateClasses = if hasRuntimeWarnings || hasRuntimeError then [ ClassName "error-tab" ] else []
+
+  -- QUESTION: what are runtimeWarnings and runtimeError? how can I reach that state?
```
best to ask @palas for some examples, these days it's quite hard to achieve an error (impossible?) or a warning (that's a good thing!)

shmish111 reviewed (changes requested):
I really like this change. I would just like you to add some more comments in places because someone coming to this afresh will need to understand both the concept and also need help with what fits into which places.

hrajchert commented:
I've updated with the comments and rebased master so we also have the latest analysis. If there are more sections that need to be commented let me know :)
